File: message-queue/haskell/src/Jmcgmqp/Runtime.hs
{-# LANGUAGE StandaloneKindSignatures, OverloadedRecordDot #-}
module Jmcgmqp.Runtime
( Instance(Instance, config, metrics, runtime)
, Runtime(..)
, newInstance
, newRuntime
, runtimeAsMap
) where

import Data.ByteString.Lazy.Char8 (pack, ByteString)
import Data.Kind (Type)
import Data.Map (Map, fromList)
import Data.Maybe (fromJust)
import Data.UUID (UUID)
import Data.UUID.V1 (nextUUID)
import Data.Time.LocalTime (getZonedTime, ZonedTime)
import Data.Time.RFC3339 (formatTimeRFC3339)
import Data.Version (showVersion)
import System.Info qualified as I
import Text.Printf (printf)

import Jmcgmqp.Config (newConfig, Config)
import Jmcgmqp.Uname (getKernel)
import Jmcgmqp.Prometheus.Metrics (newMetrics, Metrics)

type Runtime :: Type
data Runtime = Runtime {
  ctime :: ZonedTime,
  uuid :: UUID,
  lang :: String,
  lang_version :: String,
  runtime :: String,
  os :: String,
  kernel :: String,
  arch :: String
} deriving stock (Show)

newRuntime :: IO Runtime
newRuntime = do
  uuid_ <- fmap fromJust nextUUID
  kernel_ <- getKernel
  ctime_ <-  getZonedTime
  return $ Runtime {
    ctime = ctime_,
    uuid = uuid_,
    lang = "haskell",
    lang_version ="Haskell2010",
    runtime = runtime_,
    os = I.os,
    kernel = kernel_,
    arch = I.arch
    }
  where
    runtime_ :: String
    runtime_ = printf "%s %s" I.compilerName $ showVersion I.fullCompilerVersion

runtimeAsMap :: Runtime -> Map ByteString ByteString
runtimeAsMap r = fromList [
  ("ctime", pack $ formatTimeRFC3339 $ ctime r),
  ("uuid", pack $ show $ uuid r),
  ("lang", pack $ lang r),
  ("lang_version", pack $ lang_version r),
  ("runtime", pack $ r.runtime),
  ("os", pack $ os r),
  ("kernel", pack $ kernel r),
  ("arch", pack $ arch r)
  ]

type Instance :: Type
data Instance = Instance {
  config :: Config,
  runtime :: Runtime,
  metrics :: Metrics
  }

newInstance :: IO Instance
newInstance = do
  c <- newConfig
  r <- newRuntime
  Instance c r <$> newMetrics

File: message-queue/python/src/jmcgmqp/runtime.py
from dataclasses import dataclass, asdict, fields
from typing import Any
import datetime
import logging
import platform
from uuid import UUID, uuid1

from jmcgmqp.config import Config
from jmcgmqp.observer import Registry

log = logging.getLogger(__name__)

@dataclass
class Instance:
    observer: Registry = None
    runtime: "Runtime" = None
    config: "Config" = None

    def __post_init__(self):
        self.observer = Registry()
        self.runtime = Runtime.new()
        self.config = Config()
        log.info(f"Config: {asdict(self.config)}")

@dataclass(frozen=True)
class Runtime:
    ctime: datetime.datetime = None
    uuid: UUID = None
    lang: str = None
    lang_version: str = None
    runtime: str = None
    os: str = None
    kernel: str = None
    arch: str = None

    @classmethod
    def new(cls):
        uname = platform.uname()
        return cls(
            datetime.datetime.now(),
            uuid1(),
            'python',
            platform.python_version(),
            platform.python_implementation(),
            platform.system(),
            uname.release,
            uname.machine,
        )

    def metric_labels(self) -> dict[str, Any]:
        return asdict(self)

Runtime_labels = tuple(f.name for f in fields(Runtime))
